README.md
# Org-clock statistics

Compute and render Org-mode clocked time.

Check out the [demo](https://code.malloc.fr/org-clock-stat/main.html) (the initial display could appear a bit slow because random data generation is done by the browser).

Repositories are available on [GitHub](https://github.com/gscano/org-clock-stat) and [code.malloc.fr](https://code.malloc.fr/org-clock-stat).

## Usage

Displays the following __mandatory__ and ~~optional~~ CSV file columns:

__task__, __parents__, __start__, __end__, ~~effort~~, ~~ishabit~~, ~~tags~~

with every column being text except that:

* __start__ and __end__ are Javascript dates (could be YYYY-MM-DD HH:mm);
* ~~effort~~ is a Javascript duration (could be HH:mm);
* ~~ishabit~~ is equal to *t* or empty.

However, it is intended to work out of the box for *Org-mode* files processed by [org-clock-csv](https://github.com/atheriel/org-clock-csv/) which can be installed as a submodule for convenience.
Please read the [README](https://github.com/atheriel/org-clock-csv/blob/master/README.md) first in order to setup a specific init file (`~/.emacs.d/init.el` will be used by default).

On Linux, the `run.sh` script could be used to generate the targeted CSV file for a standard installation and configuration (csv content is on the standard output).
As a reminder, if no arguments are supplied and if your *Emacs* init file contains the `org-agenda-files` variable, all enclosed *Org-mode* files will be parsed.

## Install

This project depends on

* [d3.js](https://d3js.org/) version 5.12.0;
* [moment.js](https://momentjs.com/) version 2.24.0;
* [Pikaday](https://pikaday.com/) version 1.8.0.

For a better latency, run the `install.sh` command in order to retrieve and store locally the required JavaScript and CSS files. Then comment/uncomment the *remote*/*local* sections of [main.html](./main.html#L12-L24).

The general behavior can be controlled by three variables at the top of [main.js](./main.js#L1-L3) and three more at the beginning of the [`window.onload`](./main.js#L7-L9) function.

## Contributing

Contributions and ideas are most welcome, either on GitHub or by email.

ChangeLog
2020-10-22 v1.2

	* Added: include title in path.
	* Added: scroll headlines.
	* Removed: do not use categories.

2020-07-06 v1.1

	* Added: long string terminated by an ellipsis.
	* Changed: error handling and file loading for workers.
	* Fixed: start and end date detection.

2019-10-18 v1.0


test/run.sh
#!/bin/bash

../run.sh test1.org test2.org


test/test.js
var fs = require('fs');
var moment = require('moment');
var assert = require('assert');

eval(fs.readFileSync('../lib.js')+'');

function verify(lhs, rhs) {

    console.log(lhs);
    console.log(rhs);

    assert(lhs.length == rhs.length);

    assert.deepEqual(lhs, rhs);
}

/* FILTER & SPLIT */
function raw(start, end) { return {start: start, end: end}; }

var filter = createSplittingFilter({ startingDate: "2019-08-15", endingDate: "2019-08-21",
				     countWeekends: true });

verify(filter(raw("2019-08-15 13:15", "2019-08-21 11:14")),
       [raw("2019-08-15 13:15", "2019-08-21 11:14")]);

verify(filter(raw("2019-08-14 13:15", "2019-08-21 11:14")),
       [raw("2019-08-15 00:00", "2019-08-21 11:14")]);

verify(filter(raw("2019-08-15 13:15", "2019-08-22 11:14")),
       [raw("2019-08-15 13:15", "2019-08-22 00:00")]);

verify(filter(raw("2019-08-14 13:15", "2019-08-22 11:14")),
       [raw("2019-08-15 00:00", "2019-08-22 00:00")]);

var filter = createSplittingFilter({ startingDate: "2019-08-15", endingDate: "2019-08-21",
				     countWeekends: false });

verify(filter(raw("2019-08-15 13:15", "2019-08-21 11:14")),
       [raw("2019-08-15 13:15", "2019-08-17 00:00"),
	raw("2019-08-19 00:00", "2019-08-21 11:14")]);

verify(filter(raw("2019-08-14 13:15", "2019-08-21 11:14")),
       [raw("2019-08-15 00:00", "2019-08-17 00:00"),
	raw("2019-08-19 00:00", "2019-08-21 11:14")]);

verify(filter(raw("2019-08-15 13:15", "2019-08-22 11:14")),
       [raw("2019-08-15 13:15", "2019-08-17 00:00"),
	raw("2019-08-19 00:00", "2019-08-22 00:00")]);

verify(filter(raw("2019-08-14 13:15", "2019-08-22 11:14")),
       [raw("2019-08-15 00:00", "2019-08-17 00:00"),
	raw("2019-08-19 00:00", "2019-08-22 00:00")]);
/* FILTER & SPLIT */

const d1 = "2019-08-09 19:25";
const d2 = "2019-08-09 20:33";
const d3 = "2019-08-10 02:12";
const d4 = "2019-08-11 23:41";

/* DURATION */
function make(date, duration) { return [new Date(date), moment.duration(duration).asMinutes()]; }

verify(extractDaysDuration(d1, d2), [make("2019-08-09","01:08")]);
verify(extractDaysDuration(d1, d3), [make("2019-08-09","04:35"),
				     make("2019-08-10","02:12")]);
verify(extractDaysDuration(d1, d4), [make("2019-08-09","04:35"),
				     make("2019-08-10","24:00"),
				     make("2019-08-11","23:41")]);

function reduced(date, duration) { return {date: new Date(date),
					   duration: moment.duration(duration).asMinutes()}; }

function duration(start,end) { return {start:start, end:end}; }

verify(Array.from(reduceDuration([duration(d1,d2),duration(d3,d4)]))
       .map(([date, duration]) => ({date: new Date(date), duration: duration})),
       [reduced("2019-08-09", "01:08"),
	reduced("2019-08-10", "21:48"),
	reduced("2019-08-11", "23:41")]);
verify(Array.from(reduceDuration([duration(d1,d2), duration(d2,d3), duration(d3,d4)]))
       .map(([date, duration]) => ({date: new Date(date), duration: duration})),
       [reduced("2019-08-09", "04:35"),
	reduced("2019-08-10", "24:00"),
	reduced("2019-08-11", "23:41")]);
/* DURATION */

/* INTERVAL */
function interval(i,j,first,last,span) {
    var value = [];

    for(var k = i; k <= j; ++k)
	value.push(k == i ? first : (k == j ? last : span));

    return value;
}

verify(extractDaysInterval("2019-08-09 00:00", "2019-08-09 1:36", 28, Array(5).fill(0)), [28,28,28,12,0]);
verify(extractDaysInterval("2019-08-09 00:00", "2019-08-09 02:12", 35, Array(5).fill(0)), [35,35,35,27,0]);
verify(extractDaysInterval("2019-08-09 20:42", "2019-08-09 23:53", 49, Array(30).fill(0)),
       Array(25).fill(0).concat([32,49,49,49,12]));

verify(extractDaysInterval(d1, d2, 15, Array(Math.ceil(24 * 60 / 15)).fill(0)),
       Array(77).fill(0).concat(interval(77,82,5,3,15)).concat(Array(13).fill(0)));
verify(extractDaysInterval(d2, d3, 15, Array(Math.ceil(24 * 60 / 15)).fill(0)),
       interval(0,8,15,12,15).concat(Array(83 - 10).fill(0)).concat(interval(82,95,12,15,15)));
verify(extractDaysInterval(d2, d3, 28, Array(Math.ceil(24 * 60 / 28)).fill(0)),
       interval(0,4,28,20,28).concat(Array(52 - 13).fill(0)).concat(interval(44,51,27,12,28)));
verify(extractDaysInterval(d1, d3, 35, Array(Math.ceil(24 * 60 / 35)).fill(0)),
       [35,35,35,27].concat(Array(29).fill(0)).concat([25,35,35,35,35,35,35,35,5]));
/* INTERVAL */

/* LOCALE */
moment.locale('en');
assert(weekdayShift(1) == 1);

moment.locale('fr');
assert(weekdayShift(1) == 0);

moment.locale('ar');
assert(weekdayShift(1) == 2);
/* LOCALE */


run.sh
#!/bin/bash

path=$(realpath $(dirname $BASH_SOURCE))

emacs_init="~/.emacs.d/init.el"
org_clock_csv=$path"/org-clock-csv/org-clock-csv.el"

usage()
{
    echo "Usage: run [--help,-h] [--init <emacs init>] [--org-clock-csv <path>] <filename> ..."
    echo "   --init           emacs init file,     defaults to '$emacs_init'"
    echo "   --org-clock-csv  path to the library, defaults to '$org_clock_csv'"
}

parsed_opts=$(getopt -o h: -l help,org-clock-csv: -- "$@")
if [[ $? -ne 0 ]]; then usage >&2; exit 1; fi
eval "set -- $parsed_opts"
while true; do
    case "$1" in
	"-h"|"--help") usage; exit 0               ;;
	"--init") shift; emacs_init=$1             ;;
	"--org-clock-csv") shift; org_clock_csv=$1 ;;
	--) shift; break                           ;;
	*) usage >&2; exit 1                       ;;
    esac
    shift
done

emacs -batch -l "$emacs_init" -l "$org_clock_csv" -l conf.el -f org-clock-csv-batch-and-exit "$@"


data.js
// Project: [[String]: parents, String: tasks, [String]: tags, Boolean: isHabit, String: effort]

// Integer => [Project] => Integer => [String] => Integer => Integer => Float[0,1] => Integer => Float[0,1] => Integer => Float[0,1] => Integer
function randomProject(headlines, projects,
		       projectID, parents, maxDepth, maxChildren,
		       tagsProbability, maxTags,
		       effortProbability, maxEffort,
		       isHabitProbability)
{
    if(headlines == 0) return 0;

    var hasDepth = parents.length < Math.floor(Math.random() * maxDepth);
    var children = Math.min(headlines, Math.floor(Math.random() * maxChildren) + 1);

    if(!hasDepth) for(var child = 1; 0 < headlines && child <= children; ++child, --headlines) {

	var project = [];

	project.push("project " + projectID);
	project.push(parents);

	project.push(String.fromCharCode('A'.charCodeAt(0) + parents.length) + child);
	var tagsNumber = Math.floor(Math.random() * maxTags);
	var tags = [];

	while(0 < tagsNumber) {
	    if(tagsProbability < Math.random())
		tags.push("tag" + tagsNumber);
	    --tagsNumber;
	}

	project.push(tags);
	project.push(Math.random() < isHabitProbability);
	project.push((Math.random() <= effortProbability) ? displayDuration(Math.floor(Math.random() * maxEffort)) : null);

	projects.push(project);
    }
    else for(var child = 1; 0 < headlines && child <= children; ++child) {

	var localParents = [...parents];
	localParents.push(String.fromCharCode('A'.charCodeAt(0) + parents.length) + child);

	headlines = randomProject(headlines - 1, projects,
				  projectID, localParents, maxDepth, maxChildren,
				  tagsProbability, maxTags,
				  effortProbability, maxEffort,
				  isHabitProbability);
    }

    return headlines;
}

// {headlines: Integer, maxDepth: Integer, maxChildren: Integer, tagsProbability: Float[0,1], maxTags: Integer, effortProbability: Float[0,1], maxEffort: Integer, isHabitProbability: Float[0,1]} => [Project]
function randomProjects({headlines: headlines,
			 maxDepth: maxDepth, maxChildren: maxChildren,
			 tagsProbability: tagsProbability, maxTags: maxTags,
			 effortProbability: effortProbability, maxEffort: maxEffort,
			 isHabitProbability: isHabitProbability}) {

    var projects = [];
    var projectID = 1;

    while(projects.length < headlines)
	headlines = randomProject(headlines, projects,
				  projectID++, [], maxDepth, maxChildren,
				  tagsProbability, maxTags,
				  effortProbability, maxEffort,
				  isHabitProbability);

    return projects;
}

// {beforeWorkProbability: Float[0,1], morning: Integer, morningProbability: Float[0,1], lunch: Integer, lunchProbability: Float[0,1], afternoon: Integer, afternoonProbability: Float[0,1], evening: Integer, eveningProbability: Float[0,1], weekendShift: Integer} => (Moment => Boolean)
function createActivityRandomizer({beforeWorkProbability: beforeWorkProbability,
				   morning: morning, morningProbability: morningProbability,
				   lunch: lunch, lunchProbability: lunchProbability,
				   afternoon: afternoon, afternoonProbability: afternoonProbability,
				   evening: evening, eveningProbability: eveningProbability,
				   weekendShift: weekendShift}) {
    return function (date) {
	var rand = Math.random();
	if(date.isoWeekday() == 6 || date.isoWeekday() == 7) rand *= weekendShift;
	const hours = date.hours();

	if(hours < morning)
	    return [rand < beforeWorkProbability, (morning - hours) / (morning - 0)];
	else if(hours < lunch)
	    return [rand < morningProbability, (lunch - hours) / (lunch - morning)];
	else if(hours < afternoon)
	    return [rand < lunchProbability, (afternoon - hours) / (afternoon - lunch)];
	else if(hours < evening)
	    return [rand < afternoonProbability, (evening - hours) / (evening - afternoon)];
	else
	    return [rand < eveningProbability, (24 - hours) / (24 - evening)];
    }
}

// [Project] => Date => Date => (Date => Float[0,1]) => Integer => [String]
function randomData(projects, from, to, activityRandomizer, workingStep) {

    csv = ["task,parents,start,end,effort,ishabit,tags"];

    from = moment(from);
    to = moment(to);

    while(!from.isAfter(to, 'date')) {
	let [work, step] = activityRandomizer(from);

	if(work) {
	    const from_ = from.format('YYYY-MM-DD HH:mm');

	    from.add(Math.floor(Math.random() * step * workingStep) + 1, 'minutes');

	    var index = Math.floor(Math.random() * projects.length);
	    var project = projects[index];

	    var line = [];

	    line.push(project[2]);
	    line.push(project[1].slice(0, Math.min(Math.floor(Math.random() * project[1].length^2), project[1].length)).join('/'));

	    line.push(from_);
	    line.push(from.format('YYYY-MM-DD HH:mm'));

	    line.push(project[5]);
	    line.push(project[4] ? 't' : null);
	    line.push(project[3].join(':'));

	    csv.push(line.join(','));
	}
	else
	    from.add(Math.floor(Math.random() * step * workingStep) + 1, 'minutes');
    }

    return csv;
}


lib.js
function flattenHeadlines(tasks, result = {desc: [], data: []}) {
    return tasks.reduce(flattenHeadline, result);
}

function flattenHeadline({desc, data}, task) {

    desc.push({ id: task.id,
		parent: task.parentId,
		depth: task.depth,
		children: [],
		name: task.name,
		tags: task.hasOwnProperty('tags') ? task.tags : new Set(),
		effort: task.effort,
		ishabit: task.ishabit });

    data.push({ parent: task.parentId,
		entries: task.entries });

    return task.subentries.reduce(flattenHeadline, {desc, data});
}

// {startingDate: Date, endingDate: Date, countWeekends: Bool} => Bool => ({start:Date, end:Date} => [{start:Date, end:Date}])
function createSplittingFilter(config, alwaysCountWeekends = false) {
    const from = moment(config.startingDate);
    const to = moment(config.endingDate);
    const countWeekends = alwaysCountWeekends || config.hasOwnProperty('countWeekends') ? config.countWeekends : false;

    if(to.hours() == 0 && to.minutes() == 0) to.add(1, 'days').startOf('day');

    const format = 'YYYY-MM-DD HH:mm';

    function pass(date) {
	if(!countWeekends && 6 <= date.isoWeekday())
	    return [false, 8 - date.isoWeekday(), 'days'];

	return [true, 1, 'days'];
    }

    // {start:Date, end:Date} => [{start:Date, end:Date}]
    return function ({start, end}) {
	var result = [];

	start = moment(start);
	end = moment(end);

	if(start.isAfter(to) || end.isBefore(from)) return result;

	start = moment.max(start, from);
	end = moment.min(end, to);

	var recording = false;
	var current = start.clone();
	var isFirst = true;

	do {
	    const d = pass(current);

	    if(!recording && d[0]) {
		start = current.clone();
		if(!isFirst) start.startOf('day');
		recording = true;
	    }
	    else if(recording && !d[0]) {
		result.push({ start: start.format(format),
			      end: current.clone().startOf('day').format(format) });
		recording = false;
	    }

	    current.add(d[1], d[2]);
	    isFirst = false;
	}
	while(current.isBefore(end));

	if(recording)
	    result.push({ start: start.format(format), end: end.format(format) });
	else {
	    current.startOf('day');
	    const d = pass(current);
	    if(d[0] && current.isBefore(end)) {
		result.push({ start: current.format(format), end: end.format(format) });
	    }
	}

	return result;
    }
}

// Date => Date => [[Date, Int]]
function extractDaysDuration(start_, end_) {
    var start = moment(start_);
    var end = moment(end_);

    var result = [];
    const format = "YYYY-MM-DD";

    if(start.isSame(end, 'day'))
	result.push([new Date(start.format(format)), moment.duration(end.diff(start)).asMinutes()]);
    else {
	var current = start.clone().endOf('day');
	result.push([new Date(current.format(format)),
		     moment.duration(current.diff(start)).add(1, 'milliseconds').asMinutes()]);
	current.add(1, 'days').startOf('day');

	while(current.isBefore(end) && !current.isSame(end, 'day')) {
	    result.push([new Date(current.format(format)), moment.duration(1, 'day').asMinutes()]);
	    current.add(1, 'days');
	}

	result.push([new Date(current.format(format)), moment.duration(end.diff(current)).asMinutes()]);
    }

    return result;
}

// [{start:Date, end:Date}] => Map(Date => Int)
function reduceDuration(entries, result = new Map()) {

    function reduce(result, {start: start, end: end}) {

	function reduce(result, [date, duration]) {

	    var current = result.get(date.getTime());

	    if(current !== undefined)
		duration += current;

	    result.set(date.getTime(), duration);

	    return result;
	}

	return extractDaysDuration(start, end).reduce(reduce, result);
    }

    return entries.reduce(reduce, result);
}

// [{date: Date}] => {'days': Int, 'weekdays': Int}
function extractDaysInfo(data) {
    return data.reduce((accu, {date}) => {
	accu.days += 1;
	accu.weekdays += (1 <= moment(date).isoWeekday() && moment(date).isoWeekday() <= 5);

	return accu;
    },
		       ({days: 0, weekdays: 0}));
}

// => [{date: Date, duration: Int}]
function computeCalendarDurations(headlines, filter) {

    const filterFunc = createSplittingFilter(filter, true);

    const map = headlines.reduce((days, {entries}, id) =>
			       filter.headlines.has(id) ?
			       entries.reduce((days, entry) => reduceDuration(filterFunc(entry), days),
					      days)
			       : days,
			       new Map());

    const calendar = Array.from(map).map(([date, duration]) => ({date: new Date(date), duration: duration}));

    const daysCount = extractDaysInfo(calendar);

    return [calendar, daysCount];
}

// Date => Date => Int => [Int]
function extractDaysInterval(start_, end_, minutes, result) {

    const start = moment(start_);
    const end = moment(end_);

    var time = Math.floor(end.diff(start, 'minutes'));

    const firstInterval = start.hours() * 60 + start.minutes();

    var interval = Math.floor(firstInterval / minutes);
    var weight = minutes - firstInterval % minutes;

    result[interval] += weight;
    time -= weight;

    const lastInterval = 24 * 60 % minutes;

    while(minutes < time) {
	++interval;

	if(lastInterval != 0 && interval == result.length - 1) {
	    result[interval] += lastInterval;
	    time -= lastInterval;
	}
	else {
	    if(interval == result.length) interval = 0;

	    result[interval] += minutes;
	    time -= minutes;
	}
    }

    if(0 < time) {
	if(interval == result.length - 1) interval = 0;
	result[interval + 1] += time;
    }

    return result;
}

// [{start: Date, end: Date}] => Int => [Int]
function reduceInterval(entries, pace, result = Array(Math.ceil(24 * 60 / pace)).fill(0)) {
    return entries.reduce((result, {start: start, end: end}) =>
			  extractDaysInterval(start, end, pace, result),
			  result);
}

function computeDayDurations(headlines, pace, filter) {

    const filterFunc = createSplittingFilter(filter);

    return headlines.reduce((intervals,{entries},id) =>
			    filter.headlines.has(id) ?
			    entries.reduce((intervals, entry) => reduceInterval(filterFunc(entry), pace, intervals),
					   intervals)
			    : intervals,
			    Array(Math.ceil(24 * 60 / pace)).fill(0));
}

function reduceTotal(entries) {
    return entries.reduce((total, {start:start, end:end}) => moment(end).diff(start, 'minutes'), 0);
}

function computeHeadlinesDurations(headlines, filter) {
    var result = Array(headlines.length).fill().map(_ => ({total: 0, percentage: 0}));
    var total_ = 0;

    const filterFunc = createSplittingFilter(filter);

    headlines.forEach((headline, id) => {
	if(filter.headlines.has(id)) {
	    const total = headline.entries.reduce((total, entry) => total + reduceTotal(filterFunc(entry)), 0);
	    var current = id;
	    do {
		result[current].total += total;
		current = headlines[current].parent;
	    } while(current != null);

	    total_ += total;
	}
    });

    if(total_ != 0)
	headlines.forEach((headline, id) => {
	    if(headline.parent != null) {
		if(result[headline.parent].total != 0)
		    result[id].percentage = 100 * result[id].total / result[headline.parent].total;
	    }
	    else
		result[id].percentage = 100 * result[id].total / total_;
	});

    return [total_, result];
}

// Int => String([0-9][0-9])
function displayTwoDigits(number) {
    return (number < 10 ? "0" : "") + number;
}

// Int => String([0-2][0-9]:[0-5][0-9])
function displayDuration(minutes, sep = ':') {
    return displayTwoDigits(Math.floor(minutes / 60)) + sep + displayTwoDigits(minutes % 60);
}

// Interger => String
function displayLongDuration(minutes, minutesOfDay) {
    var result = "";

    const days = Math.floor(minutes / minutesOfDay);
    if(0 < days)
	result += days + "d ";

    return result + displayDuration(minutes % minutesOfDay);
}

// Int => Int
function weekdayShift(weekday) {
    const firstIsoWeekday = moment().isoWeekday(moment.weekdays(true)[0]).isoWeekday();

    if(!(weekday < firstIsoWeekday)) return 0;

    switch(firstIsoWeekday) {
    case 1: return 0;
    case 7: return 1;
    default: return 2;
    }
}

// String => 'rgb(Int(0,255),Int(0,255),Int(0,255))'
function stringToColor(str) {
    var R = 0, G = 0, B = 0;
    const variability = Math.ceil(255 / str.length);

    for(var i = 0; i < str.length; i++) {
	var code = str.charCodeAt(i);
	var value = variability * code;
	switch(i % 3) {
	case 0: R += value;
	case 1: G += value;
	case 2: B += value;
	}
    }

    R = R % 255;
    G = G % 255;
    B = B % 255;

    return `rgb(${R},${G},${B})`;
}

// String => Promise
function load(path) {
    return new Promise((resolve, reject) => {

	if(window.location.href.startsWith("file:///"))
	    return reject("Cannot load '" + path + "' (file:///).");

	var xhr = new XMLHttpRequest();

	xhr.open('GET', path, true);

	xhr.onload = function() {
	    if (xhr.status === 200)
		resolve(xhr.responseText);
	    else
		reject("Cannot load '" + path + "' (" + xhr.statusText + ")");
	}

	xhr.send();
    });
}
